iree/tools/iree-e2e-matmul-test.c

```diff
@@ -659,29 +660,37 @@ static iree_status_t replay_event_call(iree_trace_replay_t* replay,
                                     replay->host_allocator, &output_list));
   IREE_CHECK_OK(iree_vm_invoke(
       replay->context, function, IREE_VM_INVOCATION_FLAG_NONE, /*policy=*/NULL,
-      copy_of_input_list, output_list, replay->host_allocator));
+      input_list, output_list, replay->host_allocator));
 
   // Get the actual_result buffer from the output_list.
   iree_hal_buffer_view_t* actual_result;
   IREE_RETURN_IF_ERROR(
       iree_get_buffer_view_list_item(output_list, 0, &actual_result));
 
+  // Copy the results to a host local buffer to be able to map it.
+  iree_hal_buffer_view_t* host_copy_actual_result;
+  copy_buffer(replay->device, device_allocator, actual_result,
+              &host_copy_actual_result);
+
```

Comment:
I see: so far, `copy_buffer` was only called by `copy_list_of_buffer_views`, which you diff actually cements to be only used for host-to-host-only copy.

But this introduces a device-to-host use of `copy_buffer`.

In #8729 I'm renaming `copy_buffer` to `allocate_and_copy_buffer_view`.

Maybe we should further rename this to `allocate_host_buffer_view_and_copy`, making it clear that this is specialized to host destination but generic in the source (device or host).

Reply:
> I see: so far, `copy_buffer` was only called by `copy_list_of_buffer_views`, which you diff actually cements to be only used for host-to-host-only copy.

`copy_buffer` was meant to be host copy but it was a problem since the buffers sometimes would be device_local buffers. This is what I'm trying to fix.

> Maybe we should further rename this to `allocate_host_buffer_view_and_copy`, making it clear that this is specialized to host destination but generic in the source (device or host).

yes that makes sense to me. Let me rename it then. I wonder how much this will conflicts with the other patch you have?

Reply:
> > I see: so far, `copy_buffer` was only called by `copy_list_of_buffer_views`, which you diff actually cements to be only used for host-to-host-only copy.
> 
> `copy_buffer` was meant to be host copy but it was a problem since the buffers sometimes would be device_local buffers. This is what I'm trying to fix.

Sorry, i get it now: the original `input_list` is a list of device buffers, in the GPU case, so it was already performing a device-to-host copy before.

> 
> > Maybe we should further rename this to `allocate_host_buffer_view_and_copy`, making it clear that this is specialized to host destination but generic in the source (device or host).
> 
> yes that makes sense to me. Let me rename it then. I wonder how much this will conflicts with the other patch you have?

that was my next question. it conflicts quite a bit. do you want to go first? that could make sense are you are reenabling GPU test coverage, which is currently disabled. Or i can go first, but then you'll need to rerun GPU tests after you have rebased.

Reply:
> that was my next question. it conflicts quite a bit. do you want to go first? that could make sense are you are reenabling GPU test coverage, which is currently disabled. Or i can go first, but then you'll need to rerun GPU tests after you have rebased.

You should go ahead since your patch is ready to land. Does your patch changes this logic in a fundamental way or is it just going to be rebasing things? Basically what I need to change is make sure we have copies on the host of any buffers coming from `input_list`

Reply:
>  I will rebase by introducing explicit device->host copy here, perform the masking on host, and copying back to device 

Actually the masking is either overwriting the entire matrix or a no-op, so all we will need is a host->device transfer here.

Go first, then i will rebase.